src/mongoose_rabbit_worker.erl

```diff
@@ -0,0 +1,298 @@
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc
+%% Generic RabbitMQ worker. The module is responsible for interacting with
+%% RabbitMQ server. Worker holds its own AMQP connection and channel in its
+%% state.
+%% @end
+%%==============================================================================
+
+-module(mongoose_rabbit_worker).
+-author('kacper.mentel@erlang-solutions.com').
+
+-include_lib("mongooseim/include/mongoose.hrl").
+
+-behaviour(gen_server).
+
+-export([start_link/0, list_metrics/0]).
+
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2]).
+
+-type state() :: #{amqp_client_opts := mongoose_amqp:network_params(),
+                   connection := pid(),
+                   channel := pid(),
+                   host := binary(),
+                   confirms := boolean(),
+                   max_queue_len := non_neg_integer() | infinity}.
+
+-type worker_opts() :: state().
+
+-define(REPLY_REQ_DROPPED(State), {reply, request_dropped, State}).
+-define(NOREPLY_REQ_DROPPED(State), {noreply, State}).
+
+%%%===================================================================
+%%% Metrics
+%%%===================================================================
+
+-define(BACKEND, mod_event_pusher_rabbit).
+-define(CONNECTIONS_ACTIVE_METRIC, [backends, ?BACKEND, connections_active]).
+-define(CONNECTIONS_OPENED_METRIC, [backends, ?BACKEND, connections_opened]).
+-define(CONNECTIONS_CLOSED_METRIC, [backends, ?BACKEND, connections_closed]).
+-define(CONNECTIONS_FAILED_METRIC, [backends, ?BACKEND, connections_failed]).
+-define(MESSAGES_PUBLISHED_METRIC, [backends, ?BACKEND, messages_published]).
+-define(MESSAGES_FAILED_METRIC, [backends, ?BACKEND, messages_failed]).
+-define(MESSAGES_TIMEOUT_METRIC, [backends, ?BACKEND, messages_timeout]).
+-define(MESSAGE_PUBLISH_TIME_METRIC,
+        [backends, ?BACKEND, message_publish_time]).
+-define(MESSAGE_PAYLOAD_SIZE_METRIC,
+        [backends, ?BACKEND, message_payload_size]).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the RabbitMQ worker.
+%% @end
+%%--------------------------------------------------------------------
+start_link() ->
+    gen_server:start_link(?MODULE, [], []).
+
+list_metrics() ->
+    [{?CONNECTIONS_ACTIVE_METRIC, spiral},
```

Comment:
This one should be `value` or `gauge`.

Reply:
Why `spiral` is wrong here? We just increment the metric when a connection is opened and decrement when it's closed. If we use e.g. `gauge` we have to calculate somehow a total number of active connections. It seems that using `spiral` here is fairly simple solution.

Reply:
OK, right, I remembered metric type improperly. :( I had `counter` in mind. `spiral` is not *wrong* as using it's `count` property indeed works. But its feature to track value changes in 1-minute window is kind of useless here and `count` makes more sense semantically.